src/org/etools/j1939_84/controllers/part12/Part12Step12Controller.java

```diff
@@ -0,0 +1,248 @@
+                                                                     // OBD ECU
+                                                                     addFailure("6.12.12.a - NACK received for DM7 PG from OBD ECU "
+                                                                             + moduleName);
+                                                                 }
+                                                                 if (p.left.isPresent()) {
+                                                                     DM58RationalityFaultSpData packet = p.resolve();
+                                                                     // c. Fail, if expected unused bytes in DM58 are
+                                                                     // not padded with FFh
+                                                                     if (!areUnusedBytesPaddedWithFFh(packet)) {
+                                                                         addFailure("6.12.12.c - Unused bytes in DM58 are not padded with FFh in the response from "
+                                                                                 + moduleName);
+                                                                     }
+                                                                     // d. Fail, if data returned is greater than FBh
+                                                                     // (for 1 byte SP), FBFFh (for 2 byte SP), or
+                                                                     // FBFFFFFFh (for 4 byte SP)
+                                                                     if (isGreaterThanFb(packet)) {
+                                                                         addFailure("6.12.12.d - Data returned is greater than 0xFB... threshold");
+                                                                     }
+                                                                 }
+                                                             })
+                                                             .collect(Collectors.toList());
+
+                      // b. Fail, if DM58 not received (after allowed retries)
+                      if (packets.isEmpty()) {
+                          addFailure("6.12.12.d - DM58 not received (after allowed retries)");
+                      }
+
+                  });
+        });
+
+        // Actions3:
+        // a. DS DM7 with TID 245 (for DM58) using FMI 31 for first SP identified as not supporting DM58 in a DM24
+        // response In step 6.1.4.1 to the SP’s respective OBD ECU. (Use of an SP that supports test results is
+        // preferred when available).
+        // Fail/Warn criteria3:
+        // a. Fail if a NACK is not received
+        obdModules.forEach(module -> {
+            getDM58(module.getSourceAddress());
+        });
+    }
+
+    private void getDM58(int moduleAddress) {
```

Comment:
The method name is not descriptive.  This is doing more than just getting a DM58